### ISSUE TITLE
Missing SSL CA bundle on OS X

### ISSUE THREAD

**Issue by amw** [vsespb/mt-aws-glacier#87]
On OS X I wasn't able to perform HTTPS requests until I installed Mozilla::CA perl package. I was getting errors like this:

`PID 45360 HTTP connection problem (timeout?). Will retry (0 seconds spent for request)`

To learn what was really happening I had to modify GlacierRequest.pm to display `$resp->decoded_content`, which said:

```
Can't verify SSL peers without knowing which Certificate Authorities to trust

This problem can be fixed by either setting the PERL_LWP_SSL_CA_FILE
envirionment variable or by installing the Mozilla::CA module.

To disable verification of SSL peers set the PERL_LWP_SSL_VERIFY_HOSTNAME
envirionment variable to 0.  If you do this you can't be sure that you
communicate with the expected peer.
```

After a lot of googling I found out that I can download CA bundle with Mozilla::CA package. Added it to my local `cpanm` package directory and `mtglacier` started working properly. I did not had to set any additional ENV variables.

Maybe it's a good idea to add dependency on Mozilla::CA or something similar? Or at least some install notes for OS X users.

**Comment by vsespb** [comment]
Hello. Sorry you have gone through all this.

I dont have OSX, but I think README recommends to install `p5-libwww-perl` which is HTTP only. And for https there should be different package https://trac.macports.org/browser/trunk/dports/perl/p5-lwp-protocol-https/Portfile which depends on `mozilla-ca`. Anyway AFAIK any HTTPS for LWP package should depend on mozilla-ca (vendor packages and CPAN packages). 

I see you use `cpanm`. Don't you remember how did you install initial modules before this issue happened ? What modules ? all via `cpanm` ?

**Comment by amw** [comment]
I have just tried using `LWP::Protocol::https` instead of `Mozilla::CA` and it does not work. Furthermore it does not seem necessary to specify `LWP::Protocol::https` separately to run HTTPS requests. Just `LWP::UserAgent` with the addition of Mozilla's SSL certificate bundles is enough for me.

I have a bash script for running my backup. Part of it's initialization is this:

```
PERL_LIBS=./perl_libs

function install_perl {
  mkdir -p $PERL_LIBS
  curl -L http://cpanmin.us 2> /dev/null | perl - -l $PERL_LIBS \
      App::cpanminus \
      local::lib \
      LWP::UserAgent \
      JSON::XS \
      Mozilla::CA
  eval `perl -I $PERL_LIBS/lib/perl5 -Mlocal::lib=$PERL_LIBS`
}

install_perl
```

**Comment by vsespb** [comment]
Ok, I have one theory about this. What LWP version do you have ?
`perl -MLWP::UserAgent -e 'print LWP::UserAgent->VERSION'`

**Comment by vsespb** [comment]
My theroy that you have LWP::UserAgent 6.01 or 6.00.

mtglacier requires LWP::UserAgent 6+ to use HTTPS (and warns if it's below 6.xx).
so,

old versions of LWP 6.x indeed do not depend on Mozilla::CA - there is a notice in docs that
Mozilla::CA should be installed.

https://metacpan.org/pod/release/GAAS/libwww-perl-6.00/lib/LWP/UserAgent.pm
`If hostname verification is requested, and neither SSL_ca_file nor SSL_ca_path is set, then SSL_ca_file is implied to be the one provided by Mozilla::CA. If the Mozilla::CA module isn't available SSL requests will fail. Either install this module, set up an alternative SSL_ca_file or disable hostname verification.`

furthermore those versions do not use LWP::Protocol::https at all (and just ignore it).

However starting from 6.02 this notice is gone and Mozilla::CA added to dependencies

https://metacpan.org/pod/release/GAAS/libwww-perl-6.02/lib/LWP/UserAgent.pm

So, I was able to reproduce your problem `Can't verify SSL peers without knowning which Certificate Authorities to trust`
on empty Ubuntu Linux bow with the following commands:

```
curl -L http://cpanmin.us 2> /dev/null > cpanm
chmod +x cpanm
curl http://cpan.metacpan.org/authors/id/G/GA/GAAS/libwww-perl-6.00.tar.gz > libwww-perl-6.00.tar.gz
./cpanm libwww-perl-6.00.tar.gz
```

that will fail then:

```
perl -MLWP::UserAgent -e 'print LWP::UserAgent->new()->get("https://mt-aws.com")->as_string'
```

however, any attempt to install LWP::Protocol::https fixes the issue (it installs Mozilla::CA and even updates LWP::UserAgent).

So, were you using 6.01 or 6.00 ?
Even if yes, I am not sure why LWP::Protocol::https did not help.

I am going to fix several things:

1) detect missing Mozilla::CA and/or error `Can't verify SSL peers without knowning which Certificate Authorities to trust`

2) Install instructions for MacOSX do not contain even LWP::Protocol::https, so I'll add it (not sure however, what is the package name for this module and what version of LWP::Protocol is shipped with OSX 10.9, any help appreciated)

**Comment by vsespb** [comment]
oh, yes, I detect that https installed via LWP::UserAgent->is_protocol_supported("https")
but this just broken for 6.00 and 6.01

```
perl -MLWP::UserAgent -e 'print "yes" if LWP::UserAgent->is_protocol_supported("https")'
yes
```

**Comment by amw** [comment]
```
LWP::UserAgent is up to date. (6.06)
```

It did not seem to install Mozilla::CA as a dependency.

**Comment by amw** [comment]
```
photo_backup$PERL_LIBS=./perl_libs
photo_backup$curl -L http://cpanmin.us 2> /dev/null | perl - -l $PERL_LIBS \
>       App::cpanminus \
>       local::lib \
>       LWP::UserAgent \
>       JSON::XS
--> Working on App::cpanminus
Fetching http://www.cpan.org/authors/id/M/MI/MIYAGAWA/App-cpanminus-1.7004.tar.gz ... OK
Configuring App-cpanminus-1.7004 ... OK
Building and testing App-cpanminus-1.7004 ... OK
Successfully installed App-cpanminus-1.7004
--> Working on local::lib
Fetching http://www.cpan.org/authors/id/H/HA/HAARG/local-lib-2.000012.tar.gz ... OK
Configuring local-lib-2.000012 ... OK
==> Found dependencies: ExtUtils::MakeMaker
--> Working on ExtUtils::MakeMaker
Fetching http://www.cpan.org/authors/id/B/BI/BINGOS/ExtUtils-MakeMaker-6.98.tar.gz ... OK
Configuring ExtUtils-MakeMaker-6.98 ... OK
Building and testing ExtUtils-MakeMaker-6.98 ... OK
Successfully installed ExtUtils-MakeMaker-6.98 (upgraded from 6.66)
Building and testing local-lib-2.000012 ... OK
Successfully installed local-lib-2.000012
--> Working on LWP::UserAgent
Fetching http://www.cpan.org/authors/id/M/MS/MSCHILLI/libwww-perl-6.08.tar.gz ... OK
Configuring libwww-perl-6.08 ... OK
==> Found dependencies: Net::HTTP
--> Working on Net::HTTP
Fetching http://www.cpan.org/authors/id/M/MS/MSCHILLI/Net-HTTP-6.07.tar.gz ... OK
Configuring Net-HTTP-6.07 ... OK
Building and testing Net-HTTP-6.07 ... OK
Successfully installed Net-HTTP-6.07 (upgraded from 6.06)
Building and testing libwww-perl-6.08 ... OK
Successfully installed libwww-perl-6.08 (upgraded from 6.05)
--> Working on JSON::XS
Fetching http://www.cpan.org/authors/id/M/ML/MLEHMANN/JSON-XS-3.01.tar.gz ... OK
Configuring JSON-XS-3.01 ... OK
==> Found dependencies: Types::Serialiser, common::sense
--> Working on Types::Serialiser
Fetching http://www.cpan.org/authors/id/M/ML/MLEHMANN/Types-Serialiser-1.0.tar.gz ... OK
Configuring Types-Serialiser-1.0 ... OK
==> Found dependencies: common::sense
--> Working on common::sense
Fetching http://www.cpan.org/authors/id/M/ML/MLEHMANN/common-sense-3.73.tar.gz ... OK
Configuring common-sense-3.73 ... OK
Building and testing common-sense-3.73 ... OK
Successfully installed common-sense-3.73
Building and testing Types-Serialiser-1.0 ... OK
Successfully installed Types-Serialiser-1.0
Building and testing JSON-XS-3.01 ... OK
Successfully installed JSON-XS-3.01
8 distributions installed
```

**Comment by vsespb** [comment]
> It did not seem to install Mozilla::CA as a dependency.

LWP::UserAgent will _not_ install HTTPS support. LWP::Protocol::https will (only applicable to LWP::UserAgent and LWP::Protocol::https 6.02+)

**Comment by amw** [comment]
It did not seem to install Mozilla::CA either.

```
photo_backup$rm -fr perl_libs
photo_backup$curl -L http://cpanmin.us 2> /dev/null | perl - -l $PERL_LIBS \
>       App::cpanminus \
>       local::lib \
>       LWP::UserAgent \
>       LWP::Protocol::https \
>       JSON::XS
--> Working on App::cpanminus
Fetching http://www.cpan.org/authors/id/M/MI/MIYAGAWA/App-cpanminus-1.7004.tar.gz ... OK
Configuring App-cpanminus-1.7004 ... OK
Building and testing App-cpanminus-1.7004 ... OK
Successfully installed App-cpanminus-1.7004
--> Working on local::lib
Fetching http://www.cpan.org/authors/id/H/HA/HAARG/local-lib-2.000012.tar.gz ... OK
Configuring local-lib-2.000012 ... OK
==> Found dependencies: ExtUtils::MakeMaker
--> Working on ExtUtils::MakeMaker
Fetching http://www.cpan.org/authors/id/B/BI/BINGOS/ExtUtils-MakeMaker-6.98.tar.gz ... OK
Configuring ExtUtils-MakeMaker-6.98 ... OK
Building and testing ExtUtils-MakeMaker-6.98 ... OK
Successfully installed ExtUtils-MakeMaker-6.98 (upgraded from 6.66)
Building and testing local-lib-2.000012 ... OK
Successfully installed local-lib-2.000012
--> Working on LWP::UserAgent
Fetching http://www.cpan.org/authors/id/M/MS/MSCHILLI/libwww-perl-6.08.tar.gz ... OK
Configuring libwww-perl-6.08 ... OK
==> Found dependencies: Net::HTTP
--> Working on Net::HTTP
Fetching http://www.cpan.org/authors/id/M/MS/MSCHILLI/Net-HTTP-6.07.tar.gz ... OK
Configuring Net-HTTP-6.07 ... OK
Building and testing Net-HTTP-6.07 ... OK
Successfully installed Net-HTTP-6.07 (upgraded from 6.06)
Building and testing libwww-perl-6.08 ... OK
Successfully installed libwww-perl-6.08 (upgraded from 6.05)
LWP::Protocol::https is up to date. (6.04)
--> Working on JSON::XS
Fetching http://www.cpan.org/authors/id/M/ML/MLEHMANN/JSON-XS-3.01.tar.gz ... OK
Configuring JSON-XS-3.01 ... OK
==> Found dependencies: Types::Serialiser, common::sense
--> Working on Types::Serialiser
Fetching http://www.cpan.org/authors/id/M/ML/MLEHMANN/Types-Serialiser-1.0.tar.gz ... OK
Configuring Types-Serialiser-1.0 ... OK
==> Found dependencies: common::sense
--> Working on common::sense
Fetching http://www.cpan.org/authors/id/M/ML/MLEHMANN/common-sense-3.73.tar.gz ... OK
Configuring common-sense-3.73 ... OK
Building and testing common-sense-3.73 ... OK
Successfully installed common-sense-3.73
Building and testing Types-Serialiser-1.0 ... OK
Successfully installed Types-Serialiser-1.0
Building and testing JSON-XS-3.01 ... OK
Successfully installed JSON-XS-3.01
8 distributions installed
photo_backup$curl -L http://cpanmin.us 2> /dev/null | perl - -l $PERL_LIBS Mozilla::CA
--> Working on Mozilla::CA
Fetching http://www.cpan.org/authors/id/A/AB/ABH/Mozilla-CA-20130114.tar.gz ... OK
Configuring Mozilla-CA-20130114 ... OK
Building and testing Mozilla-CA-20130114 ... OK
Successfully installed Mozilla-CA-20130114
1 distribution installed
```

**Comment by vsespb** [comment]
In that log there is:
`LWP::Protocol::https is up to date. (6.04)`
that means it did not install LWP::Protocol::https because it is up to date, thus it did not install Mozilla::CA.

But how come `LWP::Protocol::https` got installed without Mozilla::CA ? - I don't know. Here is it: https://metacpan.org/source/GAAS/LWP-Protocol-https-6.04/Makefile.PL#L19 it requires Mozilla::CA

i.e. you already have installations LWP::Protocol::https, but it was broken and I am not sure why.

**Comment by vsespb** [comment]
pls check
`perl -MLWP::Protocol::https -e 'print LWP::Protocol::https->VERSION'`

**Comment by amw** [comment]
6.04

**Comment by vsespb** [comment]
Perhaps broken version shipped with Apple package, or Mozilla::CA failed just once during previous installation. Is there a link to apple dev site with source code of apple packages?

**Comment by amw** [comment]
They provide it on their open source page, but there might be some missing pieces:

https://www.opensource.apple.com

I have looked into the extras dir of perl that’s bundled with OS X and I can see that Protocol::HTTPS is here while Mozilla is not.

```
photo_backup$find /System/Library/Perl/Extras/5.18 | grep LWP | grep -i https
/System/Library/Perl/Extras/5.18/LWP/Protocol/https.pm
photo_backup$find /System/Library/Perl/Extras/5.18 | grep -i mozilla
photo_backup$
```

I don’t think it’s worthwhile to dig any further. Let’s just keep in mind we need to require Mozilla::CA specifically on OS X.

Thanks a lot for your help!

**Comment by vsespb** [comment]
Seems apple never shipped 6.04

https://trac.macports.org/changeset/122015/trunk/dports/perl/p5-lwp-protocol-https/Portfile
(is that correct link?)

6.02 changed to 6.06.

and Mozilla::CA is in dependencies: https://trac.macports.org/browser/trunk/dports/perl/p5-lwp-protocol-https/Portfile?rev=122553#L28

> I don’t think it’s worthwhile to dig any further.

ok, thank you for report!

But I can't now confirm that Mozilla::CA is required. I will only try to detect if it's missing and add runtime warnings.

feel free to re-open ticket if you have additional info.

**Comment by amw** [comment]
MacPorts is not Apple’s official packager. It’s 3rd-party packaging system that I don’t use. 6.04 is what I find on my system shipped by Apple.

**Comment by vsespb** [comment]
Ok, get it. Thanks! Well, I'll try dig a bit more by myself then.

**Comment by vsespb** [comment]
Ok, found another two reports about Apple's perl broken:

https://github.com/pjf/WebService-HabitRPG/issues/30#issuecomment-16110951
http://www.jwz.org/blog/2014/03/apple-broke-lwp-in-a-new-and-exciting-way-on-10-9-2/

will keep ticket open until fix it.

**Comment by vsespb** [comment]
Fixed it in 1.116, updated README and added runtime errors if Mozilla::CA is missing. Thanks!

**Comment by amw** [comment]
I can confirm that you are reporting the error now. Do you think it makes sense to list this as your dependency? Currently I have to:

``` bash
  curl -L http://cpanmin.us 2> /dev/null | perl - -l $PERL_LIBS \
      App::cpanminus \
      local::lib \
      LWP::UserAgent \
      JSON::XS \
      Mozilla::CA \
      App::MtAws
```

Would be nice to replace this with:

``` bash
  curl -L http://cpanmin.us 2> /dev/null | perl - -l $PERL_LIBS \
      App::cpanminus \
      local::lib \
      App::MtAws
```

And have `LWP::UserAgent`, `JSON::XS` and `Mozilla::CA` installed automatically.

**Comment by vsespb** [comment]
Hello. I added it as dependency (but did not mention it in changelog), problem that I have all HTTPS dependencies are optional.

So here it is: https://github.com/vsespb/mt-aws-glacier/blob/master/Build.PL#L171
listed as optional (`recommends` block).

HTTPS is optional cause one of mtglacier goal is ease of deploy, mtglacier distributed also as non-CPAN package (currently for Ubuntu/Debian) and can be installed without CPAN on Ubuntu/Debian 6+/RHEL5+
( example https://github.com/vsespb/mt-aws-glacier#rhelcentos-5 )
On those systems LWP::Protocol::https is anyway outdated and installing it will require installing header files/GCC/openssl dev etc. That can be important for guys who want to restore their backup in emergency or just try the app.

And in you case it is just couple of extra lines in bash script.

btw, you should probably install App::MtAws and Mozilla::CA (LWP::UserAgent and JSON::XS will be autoinstalled with App::MtAws and LWP::Protocol::https is already installed on your type of system)

**Comment by vsespb** [comment]
there should be a hack like publising another module like App::MtAws::HTTPS which will install App::MtAws and LWP::Protocol::https and Mozilla::CA, but I don't really see how it will help people who already know what CPAN is. for them it's just couple of lines in script and for me it's couple of lines in README

**Comment by amw** [comment]
Fair enough. For me HTTPS is a no brainer. If some files are important enough for me to back up to Glacier then they are important enough not to send unencrypted through the network. Frankly, backup aside, there's really few use cases for HTTP these days. You always want to at least verify identity of the other server. So many times at a coffee shop or a hotel HTTPS saved me when some application's traffic was captured by local router and redirected to their branding site. They will send you the craziest shit, redirect responses, but also 200 OK with client side redirection. Application using HTTPS will immediately know something's fishy. Application using HTTP might go crazy.

**Comment by vsespb** [comment]
Yep, I probably need to explicitly message user that he use HTTP or HTTPS. And add link to section describing all security-related stuff to WARNINGS MUST READ section of README.
btw more info related to Mozilla::CA http://blogs.perl.org/users/vsespb/2014/07/broken-lwp-in-the-wild.html
tl;dr - it's data outdated. it will be dropped some day in favour of CA file coming with OpenSSL.
